File: src/commonTest/kotlin/com/scoreo/application/CreateMatchUseCaseTest.kt
package com.scoreo.application

import com.scoreo.infrastructure.InMemoryGameTypeRepository
import com.scoreo.infrastructure.InMemoryMatchRepository
import com.scoreo.domain.model.GameType
import com.scoreo.domain.model.PlayerScore
import com.scoreo.domain.model.TieBreakRule
import com.scoreo.domain.model.WinCondition
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class CreateMatchUseCaseTest {

    private fun setup(winCondition: WinCondition): Pair<CreateMatchUseCase, InMemoryMatchRepository> {
        val gameTypeRepo = InMemoryGameTypeRepository()
        val matchRepo = InMemoryMatchRepository()
        gameTypeRepo.save(GameType("gt1", "TestGame", winCondition))
        val useCase = CreateMatchUseCase(matchRepo, gameTypeRepo)
        return Pair(useCase, matchRepo)
    }

    @Test
    fun `creates match with valid data`() {
        val (useCase, matchRepo) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase("gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)), 1767225600000L)
        assertTrue(result.isSuccess)
        assertEquals(1, matchRepo.getAll().size)
    }

    @Test
    fun `fails when game type not found`() {
        val (useCase, _) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase("unknown", listOf(PlayerScore("p1", 10)), 1767225600000L)
        assertTrue(result.isFailure)
    }

    @Test
    fun `stores manual winners when MANUAL condition`() {
        val (useCase, matchRepo) = setup(WinCondition.MANUAL)
        val result = useCase("gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)), 1767225600000L, manualWinners = listOf("p1"))
        assertTrue(result.isSuccess)
        assertEquals(listOf("p1"), matchRepo.getAll().first().manualWinners)
    }

    @Test
    fun `passes when manualWinners provided with non-MANUAL win condition (tie-break)`() {
        val (useCase, matchRepo) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase("gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 10)), 1767225600000L, manualWinners = listOf("p1"))
        assertTrue(result.isSuccess)
        assertEquals(listOf("p1"), matchRepo.getAll().first().manualWinners)
    }

    @Test
    fun `fails when manualWinners not a subset of playerScores`() {
        val (useCase, _) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase("gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)), 1767225600000L, manualWinners = listOf("unknown"))
        assertTrue(result.isFailure)
    }

    @Test
    fun `stores secondaryPlayerScores when provided`() {
        val (useCase, matchRepo) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase(
            "gt1",
            listOf(PlayerScore("p1", 10), PlayerScore("p2", 10)),
            1767225600000L,
            secondaryPlayerScores = listOf(PlayerScore("p1", 50), PlayerScore("p2", 75)),
        )
        assertTrue(result.isSuccess)
        assertEquals(2, matchRepo.getAll().first().secondaryPlayerScores.size)
    }

    @Test
    fun `fails when secondaryPlayerScores has unknown playerId`() {
        val (useCase, _) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase(
            "gt1",
            listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)),
            1767225600000L,
            secondaryPlayerScores = listOf(PlayerScore("unknown", 50)),
        )
        assertTrue(result.isFailure)
    }

    @Test
    fun `passes when manualWinners empty with non-MANUAL win condition`() {
        val (useCase, matchRepo) = setup(WinCondition.HIGHEST_SCORE)
        val result = useCase("gt1", listOf(PlayerScore("p1", 10), PlayerScore("p2", 5)), 1767225600000L)
        assertTrue(result.isSuccess)
        assertEquals(1, matchRepo.getAll().size)
    }
}

File: apps/scoreo/src/application/createMatchUseCase.test.ts
import { describe, expect, it } from 'vitest'
import type { WinCondition } from '../domain/model/enums'
import { InMemoryGameTypeRepository } from '../infrastructure/testing/inMemoryGameTypeRepository'
import { InMemoryMatchRepository } from '../infrastructure/testing/inMemoryMatchRepository'
import { CreateMatchUseCase } from './createMatchUseCase'

function setup(winCondition: WinCondition) {
  const gameTypeRepo = new InMemoryGameTypeRepository()
  const matchRepo = new InMemoryMatchRepository()
  gameTypeRepo.save({
    id: 'gt1',
    name: 'TestGame',
    winCondition,
    tieBreakRule: 'NONE',
    tieBreakCondition: 'HIGHEST_SCORE',
    tieBreakLabel: null,
    moduleId: null,
    active: true,
  })
  const useCase = new CreateMatchUseCase(matchRepo, gameTypeRepo)
  return { useCase, matchRepo }
}

describe('CreateMatchUseCase', () => {
  it('creates match with valid data', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()).toHaveLength(1)
  })

  it('fails when game type not found', () => {
    const { useCase } = setup('HIGHEST_SCORE')
    const result = useCase.invoke('unknown', [{ playerId: 'p1', score: 10 }], 1767225600000)
    expect(result.ok).toBe(false)
  })

  it('stores manual winners when MANUAL condition', () => {
    const { useCase, matchRepo } = setup('MANUAL')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
      { manualWinners: ['p1'] },
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()[0].manualWinners).toEqual(['p1'])
  })

  it('passes when manualWinners provided with non-MANUAL win condition (tie-break)', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 10 },
      ],
      1767225600000,
      { manualWinners: ['p1'] },
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()[0].manualWinners).toEqual(['p1'])
  })

  it('fails when manualWinners not a subset of playerScores', () => {
    const { useCase } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
      { manualWinners: ['unknown'] },
    )
    expect(result.ok).toBe(false)
  })

  it('stores secondaryPlayerScores when provided', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 10 },
      ],
      1767225600000,
      {
        secondaryPlayerScores: [
          { playerId: 'p1', score: 50 },
          { playerId: 'p2', score: 75 },
        ],
      },
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()[0].secondaryPlayerScores).toHaveLength(2)
  })

  it('fails when secondaryPlayerScores has unknown playerId', () => {
    const { useCase } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
      { secondaryPlayerScores: [{ playerId: 'unknown', score: 50 }] },
    )
    expect(result.ok).toBe(false)
  })

  it('stores rounds when provided', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 13 },
        { playerId: 'p2', score: 12 },
      ],
      1767225600000,
      {
        rounds: [
          [
            { playerId: 'p1', score: 10 },
            { playerId: 'p2', score: 5 },
          ],
          [
            { playerId: 'p1', score: 3 },
            { playerId: 'p2', score: 7 },
          ],
        ],
      },
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()[0].rounds).toHaveLength(2)
  })

  it('defaults rounds to an empty array when not provided', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
    )
    expect(matchRepo.getAll()[0].rounds).toEqual([])
  })

  it('passes when manualWinners empty with non-MANUAL win condition', () => {
    const { useCase, matchRepo } = setup('HIGHEST_SCORE')
    const result = useCase.invoke(
      'gt1',
      [
        { playerId: 'p1', score: 10 },
        { playerId: 'p2', score: 5 },
      ],
      1767225600000,
    )
    expect(result.ok).toBe(true)
    expect(matchRepo.getAll()).toHaveLength(1)
  })
})
